### PR TITLE
Jammy support

### PR DESCRIPTION
**Pull request by jstucke**
- add support for jammy
    - fixed installation in a lot of places
    - freetz -> freetz-ng (freetz just won't compile on jammy)
- update docker container base to jammy
- added venv support (which is also used in the docker container to fix conflicts with python system packages)

### CODE REVIEW

File: fact_extractor/helperFunctions/install.py [hunk 1]
```diff
@@ -144,26 +132,39 @@ def install_github_project(project_path: str, commands: List[str]):
...
+    """Check if FACT runs in a virtual environment"""
+    return sys.prefix != getattr(sys, 'base_prefix', getattr(sys, 'real_prefix', None))
+
+
+def gcc_is_new() -> bool:
+    try:
+        output = run(split('gcc --version'), text=True, capture_output=True).stdout
+        return int(output.splitlines()[0].split(' ')[-1].split('.')[0]) >= 10
+    except ValueError:
+        return False
```
Comment (dorpvom) [inline fact_extractor/helperFunctions/install.py:170]: This is some black magic here, hm. Why is `10` new? Should this not be parametrized?
The function's purpose would be clearer if it would go sth. like this:

```suggestion
def check_gcc_major_version_at_least(version: int = 10) -> bool:
    try:
        output = run(split('gcc --version'), text=True, capture_output=True).stdout
        major_version = int(output.splitlines()[0].split(' ')[-1].split('.')[0])
        return  major_version >= version
    except ValueError:
        return False
```

File: fact_extractor/install/common.py [hunk 2]
```diff
@@ -4,36 +4,18 @@
...
-        ]
-    },
-    'bullseye': {
-        'pip3': [
-            'testresources'
-        ]
-    },
-    # Packages common to all plateforms
+    'buster': {},
+    'bullseye': {},
```
Comment (dorpvom) [inline fact_extractor/install/common.py:17]: We will currently never get to bullseye, as it is not in the supported list in `fact_extractor/install.py`, correct? If so, we can remove all bullseye specific assignments here as well.

File: fact_extractor/plugins/unpacking/dji/test/test_plugin_dji_drones.py [hunk 3]
```diff
@@ -1,19 +1,20 @@
...
 
     def test_extraction(self):
         input_file = os.path.join(TEST_DATA_DIR, 'dji.img')
         unpacked_files, meta_data = self.unpacker.extract_files_from_file(input_file, self.tmp_dir.name)
 
-        self.assertIn("'entry_count': 2", meta_data['output'])
-        self.assertGreaterEqual(len(unpacked_files), 2, 'Should contain 2 modules')
-        self.assertIn('{}/m0305_main_controller.module'.format(self.tmp_dir.name), unpacked_files, 'At least the m0305_main_controller.module file should be matched')
+        assert "'entry_count': 2" in meta_data['output']
+        print(f"{unpacked_files=}")
```
Comment (dorpvom) [inline fact_extractor/plugins/unpacking/dji/test/test_plugin_dji_drones.py]: debug-code -> remove. You only touched this because it didn't work, ey?

File: fact_extractor/plugins/unpacking/dlink_shrs/code/dlink_shrs.py [hunk 4]
```diff
@@ -1,24 +1,24 @@
...
 
 
 def unpack_function(file_path, tmp_dir):
     decrypted_file = Path(tmp_dir, 'decrypted_image')
 
-    extraction_command = 'python3 {} -i {} -o {}'.format(TOOL_PATH, file_path, decrypted_file)
-    output = execute_shell_command(extraction_command)
-
-    return {'output': output}
+    extraction_command = f'python3 {TOOL_PATH} -i {file_path} -o {decrypted_file}'
```
Comment (dorpvom) [inline fact_extractor/plugins/unpacking/dlink_shrs/code/dlink_shrs.py:19]: Why the changes to this file?

File: fact_extractor/plugins/unpacking/intel_hex/install.sh [hunk 5]
```diff
@@ -1,5 +0,0 @@
-#!/usr/bin/env bash
-
-sudo -EH pip3 install --upgrade intelhex
```
Comment (dorpvom) [inline fact_extractor/plugins/unpacking/intel_hex/install.sh:3]: Is there a reason this setup was moved to the `install/unpacker.py`, while others weren't?

Comment (dorpvom) [reply fact_extractor/plugins/unpacking/intel_hex/install.sh:3]: Question also applies to: mikrotik, kernel, parts of dji, raw, and so on..

File: fact_extractor/plugins/unpacking/sevenz/install.sh [hunk 6]
```diff
@@ -1,22 +1,29 @@
...
 echo "      install p7z from source       "
 echo "------------------------------------"
 
-sudo -EH pip3 install --upgrade git+https://github.com/fkie-cad/common_helper_passwords.git
 
 # install newest version of p7zip
 sudo apt-get remove -y p7zip-full
 
-mkdir /tmp/fact_build
+mkdir /tmp/fact_build || :
```
Comment (dorpvom) [inline fact_extractor/plugins/unpacking/sevenz/install.sh:14]: Colon only here to send me into a 5 min shell documentation rabbit hole, hm.

File: fact_extractor/plugins/unpacking/sevenz/install.sh [hunk 7]
```diff
@@ -1,22 +1,29 @@
...
 cd /tmp/fact_build
 
 wget -O 7zip.tar.bz2 https://sourceforge.net/projects/p7zip/files/latest/download
 tar xvjf 7zip.tar.bz2
-(cd p7zip* && cp makefile.linux_amd64_asm makefile.linux && make -j$(nproc) all3 && sudo ./install.sh)
+cd p7zip*
+# gcc >= 11 has -Wnarrowing as default flag which leads to an error during compilation
+# g++ will try to use standard C++17 but the code is not compatible -> use C++14
+sed -i 's/CXXFLAGS=-c -I. \\/CXXFLAGS=-c -I. -Wno-narrowing -std=c++14 \\/g' makefile.glb  || echo "Warning: Could not apply makefile patch"
```
Comment (dorpvom) [inline fact_extractor/plugins/unpacking/sevenz/install.sh:22]: Good thing we have tests...

File: fact_extractor/plugins/unpacking/uefi/install.sh [hunk 8]
```diff
@@ -1,21 +0,0 @@
...
-
-cd "$( dirname "${BASH_SOURCE[0]}" )"
-
-echo "------------------------------------"
-echo "        install uefi parser         "
-echo "------------------------------------"
-
-cd ../../../install
-
-git clone https://github.com/theopolis/uefi-firmware-parser.git
```
Comment (dorpvom) [inline fact_extractor/plugins/unpacking/uefi/install.sh:11]: So this was resolved in the meantime?

**Review by dorpvom — CHANGES_REQUESTED** [review]
Please consider or comment on the annotations

File: Dockerfile [hunk 9]
```diff
@@ -7,9 +7,16 @@ COPY . /opt/app
 WORKDIR /opt/app/fact_extractor
 
 ARG USER=root
+ARG DEBIAN_FRONTEND=noninteractive
 
-RUN install/pre_install.sh
+RUN install_clean python3.11 python3.11-dev python3.11-venv gcc
```
[the remark this answers is not in the document]
Comment (jstucke) [reply Dockerfile:12]: Originally I added this to fix compatibility issues with existing apt python packages. Since I also fixed it for the CI afterwards (without venv) it could be no longer necessary. It should be generally more robust, though, so we might want to keep it.

File: fact_extractor/plugins/unpacking/dlink_shrs/code/dlink_shrs.py (hunk 4 above)
Comment (jstucke) [reply fact_extractor/plugins/unpacking/dlink_shrs/code/dlink_shrs.py:19]: There were problems with the tests and I added some debugging code which I then removed after it worked again. So the changes are no longer needed. I can restore the original version if you want.

File: fact_extractor/install/unpacker.py [hunk 10]
```diff
@@ -114,29 +155,67 @@
...
+            'make',
+            'patch',
+            'perl',
             'pkg-config',
+            'pv',
+            'rsync',
+            'sqlite3',
+            'subversion',
+            'texinfo',
+            'tofrodos',
```
[the remark this answers is not in the document]
Comment (jstucke) [reply fact_extractor/install/unpacker.py:210]: Those were some dependencies I copied from https://freetz-ng.github.io/freetz-ng/PREREQUISITES/ but I'm actually not really sure what they do :sweat_smile:

File: fact_extractor/plugins/unpacking/intel_hex/install.sh (hunk 5 above)
Comment (jstucke) [reply fact_extractor/plugins/unpacking/intel_hex/install.sh:3]: I only moved python package installation to `install/unpacker.py` to be able to adjust the installation in case of a venv without having to do this in every single `install.sh` (and I also didn't want to convert every installation to a `install.py` but maybe this is a thing that we should do at some point :sweat_smile: )

File: fact_extractor/plugins/unpacking/sevenz/install.sh (hunk 6 above)
Comment (jstucke) [reply fact_extractor/plugins/unpacking/sevenz/install.sh:14]: I googled shell script nop and `:` was the result :sweat_smile:

Comment (jstucke) [reply fact_extractor/plugins/unpacking/sevenz/install.sh:14]: The idea was to not fail if the directory already exists (in case the 7z installation failed halfway through) but I will change it to a `-p` to make it clearer

File: fact_extractor/plugins/unpacking/sevenz/install.sh (hunk 7 above)
Comment (jstucke) [reply fact_extractor/plugins/unpacking/sevenz/install.sh:22]: It was really fun debugging this :facepalm:

File: fact_extractor/plugins/unpacking/uefi/install.sh (hunk 8 above)
Comment (jstucke) [reply fact_extractor/plugins/unpacking/uefi/install.sh:11]: Yes, apparently there were some major updates to `uefi-firmware-parser` also with new packages on PyPI which makes the installation much easier

File: fact_extractor/plugins/unpacking/sevenz/install.sh (hunk 6 above)
Comment (dorpvom) [reply fact_extractor/plugins/unpacking/sevenz/install.sh:14]: Nah, it's okay. I'm just used to `|| true` for those purposes and was not aware of what `:` does here.

File: fact_extractor/plugins/unpacking/sevenz/install.sh (hunk 7 above)
Comment (jstucke) [reply fact_extractor/plugins/unpacking/sevenz/install.sh:22]: The question is: should we pin the 7z version so that this will not cause a problem with a new version?

File: fact_extractor/install/unpacker.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dorpvom) [reply fact_extractor/install/unpacker.py:210]: FTR tofrodos is short for tofromdos or to/from dos and in my understanding is only used to convert between CR and CR/LF.

File: fact_extractor/plugins/unpacking/dji/test/test_plugin_dji_drones.py (hunk 3 above)
Comment (jstucke) [reply fact_extractor/plugins/unpacking/dji/test/test_plugin_dji_drones.py]: removed

File: fact_extractor/helperFunctions/install.py (hunk 1 above)
Comment (jstucke) [reply fact_extractor/helperFunctions/install.py:170]: > Why is 10 new?
AFAIK 10 was the version where changes to default warning flags occurred (which can lead to build errors with `-Werror`)

> Should this not be parametrized?
Yes, good idea!

Comment (jstucke) [reply fact_extractor/helperFunctions/install.py:170]: I also just found out that there is a `gcc -dumpversion` option which should make this less ugly.

File: fact_extractor/install.py [hunk 11]
```diff
@@ -40,7 +40,8 @@
 # Compatible Ubuntu releases
 BIONIC_CODE_NAMES = ['bionic', 'tara', 'tessa', 'tina', 'disco']
 XENIAL_CODE_NAMES = ['xenial', 'yakkety', 'sarah', 'serena', 'sonya', 'sylvia']
```
[the remark this answers is not in the document]
Comment (jstucke) [reply fact_extractor/install.py:42]: Removed everything xenial-related

File: fact_extractor/install/common.py (hunk 2 above)
Comment (jstucke) [reply fact_extractor/install/common.py:17]: actually it is included in `fact_extractor/install.py`. I also added it in the exit message. Also why remove bullseye and keep buster when bullseye is newer?

Comment (jstucke) [reply fact_extractor/install/common.py:17]: There was also some rudimentary support for Kali but it was included in the buster codenames. I changed it to bullseye (which makes a bit more sense IMHO as it's base is probably newer than buster)

File: fact_extractor/plugins/unpacking/uefi/install.sh (hunk 8 above)
Comment (dorpvom) [reply fact_extractor/plugins/unpacking/uefi/install.sh:11]: I can expand on this: Though pypi lists the package on version 1.10 and there is a correspondin release from January 5th, when installing with pip it does not recognize 1.10 and lists 1.9 as latest version. That version is not running on python3 seemingly (details, details, nevermind) so that the plugin also can not work correctly.

A possible fix would be to install `git+https://github.com/theopolis/uefi-firmware-parser@v1.10`.

Comment (jstucke) [reply fact_extractor/plugins/unpacking/uefi/install.sh:11]: That is interesting. If I remember correctly, the test failed for me, too, but just once. After that I could not reproduce it anymore. There could also be a problem with the python version as https://pypi.org/project/uefi-firmware/ lists the package to be python 3.6/7 compatible but no newer versions. I will change it to `git+https://github.com/theopolis/uefi-firmware-parser@v1.10`. At least it seems to install without issues in the container that way.

File: fact_extractor/plugins/unpacking/intel_hex/install.sh (hunk 5 above)
Comment (dorpvom) [reply fact_extractor/plugins/unpacking/intel_hex/install.sh:3]: I do understand that, it's just one more thing that binds plugins to the _core_. I'd say if we go down this road we can drop the plugin `install.sh` files altogether, but I guess for the moment we can keep it like this and ignore the ambiguity...

File: fact_extractor/helperFunctions/install.py (hunk 1 above)
Comment (jstucke) [reply fact_extractor/helperFunctions/install.py]: FTR: switching to a [newer Fork of sasquatch](https://github.com/threadexio/sasquatch) made this unnecessary so I removed it completely

File: fact_extractor/plugins/unpacking/intel_hex/install.sh (hunk 5 above)
Comment (jstucke) [reply fact_extractor/plugins/unpacking/intel_hex/install.sh:3]: I would prefer a solution closer to what we did in FACT_core (setup.py/requirements.txt/apt_packages.txt/etc.) but this is absolutely out of scope for this PR. I don't know how relevant custom plugins are for the extractor, though.

**Review by dorpvom — APPROVED** [review]
